File: app/jms/command.py
import re
import time
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel
from starlette.websockets import WebSocket

from jms.base import BaseWisp
from api.schemas import AskResponse, AskResponseType
from wisp.protobuf import service_pb2
from wisp.exceptions import WispError
from wisp.protobuf.common_pb2 import Session, CommandACL, RiskLevel
from utils import reply
from utils.logger import get_logger

logger = get_logger(__name__)


class CommandRecord(BaseModel):
    input: Optional[str] = None
    output: Optional[str] = None
    risk_level: str = RiskLevel.Normal


class CommandHandler(BaseWisp):
    REJECT_MESSAGE = "reject by acl rule"
    WAIT_TICKET_TIMEOUT = 60 * 3
    WAIT_TICKET_INTERVAL = 3

    def __init__(self, websocket: WebSocket, session: Session, command_acls: List[CommandACL]):
        super().__init__()
        self.session = session
        self.websocket = websocket
        self.command_acls = command_acls

    async def record_command(self, command_record: CommandRecord):
        req = service_pb2.CommandRequest(
            sid=self.session.id,
            org_id=self.session.org_id,
            asset=self.session.asset,
            account=self.session.account,
            user=self.session.user,
            timestamp=int(datetime.timestamp(datetime.now())),
            input=command_record.input,
            output=command_record.output,
            risk_level=command_record.risk_level
        )
        resp = self.stub.UploadCommand(req)
        if not resp.status.ok:
            error_message = f'Failed to upload command: {resp.status.err}'
            logger.error(error_message)
            raise WispError(error_message)

    def match_rule(self, command: str):
        for command_acl in self.command_acls:
            for command_group in command_acl.command_groups:
                flags = re.UNICODE
                if command_group.ignore_case:
                    flags |= re.IGNORECASE
                try:
                    pattern = re.compile(command_group.pattern, flags)
                    if pattern.search(command.lower()) is not None:
                        return command_acl
                except re.error as e:
                    error_message = f'Failed to re invalid pattern: {command_group.pattern} {e}'
                    logger.error(error_message)
                    # TODO Exception
                    raise Exception(error_message)

    async def create_and_wait_ticket(self, command: str, command_acl: CommandACL) -> bool:
        req = service_pb2.CommandConfirmRequest(
            cmd=command,
            session_id=self.session.id,
            cmd_acl_id=command_acl.id
        )
        resp = self.stub.CreateCommandTicket(req)
        if not resp.status.ok:
            error_message = f'Failed to create ticket: {resp.status.err}'
            logger.error(error_message)
            raise WispError(error_message)

        return await self.wait_for_ticket_status_change(resp.info)

    async def wait_for_ticket_status_change(self, ticket_info: service_pb2.TicketInfo):
        await reply(
            self.websocket, AskResponse(
                type=AskResponseType.waiting,
                conversation_id=self.session.id,
                system_message=f'等待工单审批: {ticket_info.ticket_detail_url}'
            )
        )
        start_time = time.time()
        end_time = start_time + self.WAIT_TICKET_TIMEOUT

        ticket_closed = False
        is_continue = False
        while time.time() <= end_time:
            check_request = service_pb2.TicketRequest(req=ticket_info.check_req)
            check_response: service_pb2.TicketStateResponse = self.stub.CheckTicketState(check_request)

            if not check_response.status.ok:
                error_message = f'Failed to check ticket status: {check_response.status.err}'
                logger.error(error_message)
                break

            state = check_response.Data.state
            if state == service_pb2.TicketState.Approved:
                is_continue = True
                ticket_closed = True
                break
            elif state in [service_pb2.TicketState.Rejected, service_pb2.TicketState.Closed]:
                ticket_closed = True
                await reply(
                    self.websocket, AskResponse(
                        type=AskResponseType.waiting,
                        conversation_id=self.session.id,
                        system_message=f'工单关闭或拒绝'
                    )
                )
                break

            time.sleep(self.WAIT_TICKET_INTERVAL)

        if not ticket_closed:
            self.close_ticket(ticket_info)

        return is_continue

    async def command_acl_filter(self, command: CommandRecord):
        is_continue = False
        acl = self.match_rule(command.input)
        if acl is not None:
            command.risk_level = RiskLevel.Danger
            if acl.action == CommandACL.Reject:
                await reply(
                    self.websocket, AskResponse(
                        type=AskResponseType.reject,
                        conversation_id=self.session.id,
                        system_message=self.REJECT_MESSAGE
                    )
                )
            elif acl.action == CommandACL.Review:
                try:
                    is_continue = self.create_and_wait_ticket(command.input, acl)
                except Exception as e:
                    print(command.input, str(e))
            else:
                is_continue = True
        return is_continue

    def close_ticket(self, ticket_info: service_pb2.TicketInfo):
        req = service_pb2.TicketRequest(req=ticket_info.cancel_req)
        resp = self.stub.CancelTicket(req)
        if not resp.status.ok:
            error_message = f'Failed to close ticket: {resp.status.err}'
            logger.error(error_message)

File: pkg/jms/command.go
package jms

import (
	"context"
	"encoding/json"
	"fmt"
	"github.com/dlclark/regexp2"
	"github.com/gorilla/websocket"
	"github.com/jumpserver/kael/pkg/httpd/grpc"
	"github.com/jumpserver/kael/pkg/logger"
	"github.com/jumpserver/kael/pkg/schemas"
	"github.com/jumpserver/wisp/protobuf-go/protobuf"
	"time"
)

type CommandHandler struct {
	Session       *protobuf.Session
	Websocket     *websocket.Conn
	CommandACLs   []*protobuf.CommandACL
	CmdACLID      string
	CmdGroupID    string
	CommandRecord *schemas.CommandRecord
	JMSState      *schemas.JMSState
}

const (
	WAIT_TICKET_TIMEOUT  = 60 * 3
	WAIT_TICKET_INTERVAL = 2
)

func NewCommandHandler(
	websocket *websocket.Conn, session *protobuf.Session,
	commandACLs []*protobuf.CommandACL, jmsState *schemas.JMSState,
) *CommandHandler {
	return &CommandHandler{
		Session:       session,
		Websocket:     websocket,
		CommandACLs:   commandACLs,
		CommandRecord: nil,
		JMSState:      jmsState,
	}
}

func (ch *CommandHandler) RecordCommand() {
	ctx := context.Background()
	req := &protobuf.CommandRequest{
		Sid:        ch.Session.Id,
		OrgId:      ch.Session.OrgId,
		Asset:      ch.Session.Asset,
		Account:    ch.Session.Account,
		User:       ch.Session.User,
		Timestamp:  time.Now().Unix(),
		Input:      ch.CommandRecord.Input,
		Output:     ch.CommandRecord.Output,
		RiskLevel:  ch.CommandRecord.RiskLevel,
		CmdAclId:   ch.CmdACLID,
		CmdGroupId: ch.CmdGroupID,
	}

	resp, _ := grpc.GlobalGrpcClient.Client.UploadCommand(ctx, req)
	if !resp.Status.Ok {
		logger.GlobalLogger.Error("Failed to upload command")
	}
}

func (ch *CommandHandler) MatchRule() *protobuf.CommandACL {
	for _, commandACL := range ch.CommandACLs {
		for _, commandGroup := range commandACL.CommandGroups {
			regexp2Opt := regexp2.None
			if commandGroup.IgnoreCase {
				regexp2Opt = regexp2.IgnoreCase
			}

			re, err := regexp2.Compile(commandGroup.Pattern, regexp2Opt)
			if err != nil {
				logger.GlobalLogger.Error("Failed to compile regular expression")
				return nil
			}

			found, err := re.FindStringMatch(ch.CommandRecord.Input)
			if err != nil || found == nil {
				continue
			}

			ch.CmdACLID = commandACL.Id
			ch.CmdGroupID = commandGroup.Id
			return commandACL
		}
	}
	return nil
}

func (ch *CommandHandler) CreateAndWaitTicket(commandACL *protobuf.CommandACL) bool {
	ctx := context.Background()
	req := &protobuf.CommandConfirmRequest{
		Cmd:       ch.CommandRecord.Input,
		SessionId: ch.Session.Id,
		CmdAclId:  ch.CmdACLID,
	}

	resp, _ := grpc.GlobalGrpcClient.Client.CreateCommandTicket(ctx, req)
	if !resp.Status.Ok {
		logger.GlobalLogger.Error("Failed to create ticket")
		return false
	}

	return ch.WaitForTicketStatusChange(resp.Info)
}

func (ch *CommandHandler) WaitForTicketStatusChange(ticketInfo *protobuf.TicketInfo) bool {
	response := &schemas.AskResponse{
		Type:           schemas.Waiting,
		ConversationID: ch.Session.Id,
		SystemMessage: fmt.Sprintf(
			"复核请求已发起，请等待复核: [查看结果](%s): ",
			ticketInfo.TicketDetailUrl,
		),
	}

	jsonResponse, _ := json.Marshal(response)
	_ = ch.Websocket.WriteMessage(websocket.TextMessage, jsonResponse)

	ctx := context.Background()
	startTime := time.Now()
	endTime := startTime.Add(time.Duration(WAIT_TICKET_TIMEOUT) * time.Second)
	ticketClosed := true
	isContinue := false
	for time.Now().Before(endTime) {
		req := &protobuf.TicketRequest{Req: ticketInfo.CheckReq}
		resp, _ := grpc.GlobalGrpcClient.Client.CheckTicketState(ctx, req)
		if !resp.Status.Ok {
			logger.GlobalLogger.Error("Failed to check ticket status")
			break
		}
		switch resp.Data.State {
		case protobuf.TicketState_Approved:
			isContinue = true
			ticketClosed = false
			ch.CommandRecord.RiskLevel = protobuf.RiskLevel_ReviewAccept
			break
		case protobuf.TicketState_Rejected:
			ch.CommandRecord.RiskLevel = protobuf.RiskLevel_ReviewReject
			ticketClosed = false

		case protobuf.TicketState_Closed:
			ch.CommandRecord.RiskLevel = protobuf.RiskLevel_ReviewCancel
			ticketClosed = false

		default:
			time.Sleep(WAIT_TICKET_INTERVAL)
		}
	}

	if ticketClosed {
		ch.CloseTicket(ticketInfo)
	}
	return isContinue
}

func (ch *CommandHandler) CommandACLFilter() bool {
	isContinue := true
	acl := ch.MatchRule()
	if acl != nil {
		switch acl.Action {
		case protobuf.CommandACL_Reject:
			response := &schemas.AskResponse{
				Type:           schemas.Reject,
				ConversationID: ch.Session.Id,
				SystemMessage:  "对话已被拒绝",
			}

			jsonResponse, _ := json.Marshal(response)
			_ = ch.Websocket.WriteMessage(websocket.TextMessage, jsonResponse)

			isContinue = false
			ch.CommandRecord.RiskLevel = protobuf.RiskLevel(protobuf.CommandACL_Reject)
		case protobuf.CommandACL_Warning:
			ch.CommandRecord.RiskLevel = protobuf.RiskLevel(protobuf.CommandACL_Warning)
		case protobuf.CommandACL_Review:
			response := &schemas.AskResponse{
				Type:           schemas.Waiting,
				ConversationID: ch.Session.Id,
				SystemMessage:  "您输入命令需要复核后才可以执行，是否发起复核请求？",
				Meta:           schemas.ResponseMeta{ActivateReview: true},
			}

			jsonResponse, _ := json.Marshal(response)
			_ = ch.Websocket.WriteMessage(websocket.TextMessage, jsonResponse)

			isContinue = false
			startTime := time.Now()
			endTime := startTime.Add(time.Duration(60) * time.Second)

			for time.Now().Before(endTime) {
				switch ch.JMSState.ActivateReview {
				case schemas.Wait:
					time.Sleep(1 * time.Second)
				case schemas.Rejected:
					isContinue = false
					ch.JMSState.ActivateReview = schemas.Wait
					break
				case schemas.Approve:
					ch.JMSState.ActivateReview = schemas.Wait
					isContinue = ch.CreateAndWaitTicket(acl)
					break
				}
			}
		}
	}
	return isContinue
}

func (ch *CommandHandler) CloseTicket(ticketInfo *protobuf.TicketInfo) {
	ctx := context.Background()
	req := &protobuf.TicketRequest{
		Req: ticketInfo.CancelReq,
	}

	resp, _ := grpc.GlobalGrpcClient.Client.CancelTicket(ctx, req)
	if !resp.Status.Ok {
		logger.GlobalLogger.Error("Failed to close ticket")
	}
}
